### DISCUSSION TITLE
Set or Fill using a factory when testing Livewire Component

### DISCUSSION THREAD

**Discussion by chrisrhymes** [livewire/livewire#3892] [1. Help]
How can you use a factory in your tests so set all the attributes? Below is an example component and a test. I'm trying to set the values of the $user from the factory. I know I can set the attributes one by one using dot notation like `user.name`, but I have models where there are many attributes to set and using a factory is much easier. 

If I run the below I get the following error
`Livewire\Exceptions\CannotBindToModelDataWithoutValidationRuleException : Cannot bind property [user] without a validation rule present in the [$rules] array on Livewire component: [example].`

```php
// An example component
<?php

namespace App\Http\Livewire;

use App\Models\User;
use Livewire\Component;

class Example extends Component
{
    public User $user;

    public function mount()
    {
        $this->user = new User();
    }

    public function rules()
    {
        return [
            'user.name' => ['required', 'max:200'],
            'user.email' => ['required', 'email'],
        ];
    }

    public function save()
    {
        $this->validate();

        $this->user->save();
    }

    public function render()
    {
        return view('livewire.example');
    }
}
```

```php
// An example test
<?php

namespace Tests\Feature;

use App\Http\Livewire\Example;
use App\Models\User;
use Illuminate\Foundation\Testing\RefreshDatabase;
use Livewire\Livewire;
use Tests\TestCase;

class ExampleTest extends TestCase
{
    use RefreshDatabase;

    public function test()
    {
        $user = User::factory()->create();

        Livewire::test(Example::class)
            ->set('user', $user)
            ->set('user.email', 'fake@example.com')
            ->set('user.name', 'Joe Bloggs')
            ->call('save')
            ->assertHasNoErrors();
    }
}
```

If I then add `'user' => ''` to the rules array in the Example component it allows me to set the attributes to the user model, but it then has the following error message:
`Illuminate\Database\QueryException : SQLSTATE[HY000]: General error: 1 no such column: user (SQL: update "users" set "name" = Joe Bloggs, "email" = fake@example.com, "user" = ?, "updated_at" = 2021-09-28 10:58:28 where "id" = 1)`

It seems it is trying to insert the user model as a column in user? 

I have also tried setting `'user.*' => ''` to the rules array and now the test passes, but when I add relations to the User it converts the relation to an array and it appears in the attributes array of the user instead of the relations. This then fails to save as it looks for the relation name as a column in the users table. Thanks.

**Comment by ryatkins** [comment]
Try not setting 'user' to $user and just set the properties you have defined.

```
    public function test()
    {
        $user = User::factory()->create();

        Livewire::test(Example::class)
            ->set('user.email', $user->email)
            ->set('user.name', $user->name)
            ->call('save')
            ->assertHasNoErrors();
    }
```

**Reply by chrisrhymes** [reply]
Thanks @ryatkins that will work, but I have some models/Livewire forms that are very large and it's just a bit tedious doing that one by one when really I want to write a quick feature test to test that notifications are sent on form submission, or that a job is queued or an event is triggered, etc. rather than testing the form itself.

**Reply by jonneroelofs** [reply]
you could do something like
```php
public function mount(User $user){
    $this->user = $user ?? User::make()
}
```

and then in your test mount the user:
```php
Livewire::test(Example::class, ['user' => $user])
```

**Comment by chrisrhymes** [comment]
Thanks for the comments and advice. For now I created a custom mixin to allow me to do what I wanted and there is a PR here with something similar. https://github.com/livewire/livewire/pull/3901. 

```php
<?php

namespace Tests\Mixins;

use Closure;

/**
 * @mixin \Livewire\Testing\TestableLivewire
 */
class CustomLivewireTestMixin
{
    public function fillFactory(): Closure
    {
        return function ($factory, string $prefix, array $except = []) {
            collect($factory)
                ->except($except)
                ->each(function ($value, $field) use ($prefix) {
                    if ($value !== null) {
                        $this->set("$prefix.$field", $value);
                    }
                });

            return $this;
        };
    }
}
```

So in the test I can do this (I actually registered the mixin in the TestCase in my project, but this is just an example). 

```php
<?php

namespace Tests\Feature;

use App\Http\Livewire\Example;
use App\Models\User;
use Illuminate\Foundation\Testing\RefreshDatabase;
use Livewire\Livewire;
use Tests\TestCase;
use Livewire\Testing\TestableLivewire;
use Tests\Mixins\CustomLivewireTestMixin;

class ExampleTest extends TestCase
{
    use RefreshDatabase;

    public function test()
    {
        TestableLivewire::mixin(new CustomLivewireTestMixin());

        $user = User::factory()->create();

        Livewire::test(Example::class)
            ->fillFactory($user, 'user', ['password'])
            ->call('save')
            ->assertHasNoErrors();
    }
}
```

**Reply by ryatkins** [reply]
Very nice!
